### ISSUE TITLE
Having trouble getting sdk working with require.js (AWS class not in scope)

### ISSUE THREAD

**Issue by scottw-finao** [aws/aws-sdk-js#3017]
I'm trying to figure out how to use the browser-based aws-sdk.js with require.js when building a web app.

If I try to build a class and include aws-sdk in the require.js define, once I try to create an instance and reference the AWS class, it says it is undefined.

For example, I have a function that checks to see if the credentials are an instance of AWS.Credentials and if not, it tries to initialize it with an STS token it retrieves via Rest. The problem is, the code dies on the if(this.credentials instanceof AWS.Credentials) saying AWS is undefined. It even dies on my simple check of the needsRefresh.

This is what my define looks like - I'll include the 'needsRefresh' wrapper for an example of the sort of thing that is throwing the Undefined error:

```
define(['require','aws-sdk','jquery'],
    function (require, AWS, $) {

        // Aws helper class
        function AwsHelper() { };

        AwsHelper.prototype = {
            credentials: null,

            tokenNeedsRefresh: function () {
                //////////////////////////////////////////////////////////////////
                // errors out on the following line with:                       //
                //   TypeError: Cannot read property 'Credentials' of undefined //
                //////////////////////////////////////////////////////////////////
                if(this.credentials instanceof AWS.Credentials) {
                    return this.credentials.needsRefresh();
                } else return true;
            }
        };
        return AwsHelper;
    }
);
```

I also tried the following format at the top of the file:

```
define(function (require) {
        var AWS = require("aws-sdk"),
            $ = require("jquery");
        /* .. */

```
If I remove all onLoad references to the refresh code running, it will load and I can create an instance. But as soon as I call any function that references the AWS class, it dies.

How do I make sure that AWS class definition is still in global space once an instance is spawned?

related open thread:

unresolved [Stack overflow](https://stackoverflow.com/questions/59330857/using-aws-sdk-js-with-require-js-in-a-browser)

unanswered [Require.js github](https://github.com/requirejs/requirejs/issues/1798)

unanswered [AWS community forum](https://forums.aws.amazon.com/thread.jspa?messageID=926531)

**Comment by ajredniwja** [comment]
Hey @scottw-finao, thank-you for reaching out to us with your issue. 

I believe the error is not because of the SDK. 

If you are running it on a browser you need to add a script tag as mentioned [here](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/getting-started-browser.html#getting-started-browser-run-sample).

Can you also try running it without requireJS to just isolate where the problem is coming from since I am not able to reproduce it when running locally.

**Comment by scottw-finao** [comment]
I did try that also - the problem is that the sdk still isn't loaded when the require.js code runs. I even tried adding a delay/retry scheme using a setTimeout recursive check, but for some reason, it will see AWS defined but then it's still undefined when the function is called.  I'm not sure what all require.js or the aws-sdk.js is doing so I'm not entirely sure what is going on.
Most of the AWS examples show setting static-public values to set the configuration, but I can't even seem to get new instances formed when I need them to then set the values per instance.

**Comment by scottw-finao** [comment]
(the idea of idea.js configuration is it's essentially an asynchronous, lazy-loaded scheme. The libraries are supposed to be loaded before any code runs. But something in how aws-sdk.js initializes doesn't work with require.js/commonjs setup)

**Comment by ajredniwja** [comment]
Hey @scottw-finao, I believe at that point the error doesn't come from the SDK itself but something with from the requireJS. I can try to find a workaround but not sure about it.

**Comment by scottw-finao** [comment]
I've side-lined that attempt for the moment. I may re-address it trying other strategies such as separating the SDK stuff and holding off all references to AWS until it's properly loaded. My main reason for looking at it was a need to use a utility library more than once, but since there were only one or two utility functions being used in the multiple locations, I can just put them in multiple places.
I just prefer to do them in one - less maintenance nightmare when updating a core util function.

**Comment by scottw-finao** [comment]
any updates on this? While I'm doing in in-line loading methods via the html, it would still be preferable to use require.js

**Comment by ajredniwja** [comment]
@scottw-finao, I was not able to work around it since it uses a third party library, which fails to load the SDK.

**Comment by scottw-finao** [comment]
I managed to get it working yesterday a couple different ways. First, I noted that requirejs was creating a local scope AWS object that was empty when I added it to the array and then also added it as 'AWS' in the function parameters of the define. But what I hadn't noticed is that AWS was also defined in the global scope, so removing the 'AWS' in the function allowed me to access the class.
I then started playing around with shim configurations in relation to some other non AMD libraries, and found that I could define a shim with "exports":"AWS" and it also works.
